pytorch_lightning/plugins/training_type/training_type_plugin.py

```diff
@@ -0,0 +1,121 @@
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+import os
+from abc import ABC, abstractmethod
+from typing import Any, Optional, Sequence, Union
+
+import torch
+
+from pytorch_lightning import _logger as log
+from pytorch_lightning.core.lightning import LightningModule
+from pytorch_lightning.plugins.base_plugin import Plugin
+from pytorch_lightning.trainer import Trainer
+
+
+class TrainingTypePlugin(Plugin, ABC):
+    """A Plugin to change the behaviour of the training, validation and test-loop."""
+
+    def __init__(self) -> None:
+        self._model = None
+        self._results = None
+        self.global_rank = 0
+
+    @property
+    @abstractmethod
+    def on_gpu(self) -> bool:
+        """Returns whether the current process is done on GPU"""
+        raise NotImplementedError
+
+    @property
+    @abstractmethod
+    def root_device(self) -> torch.device:
+        """Returns the root device"""
+        raise NotImplementedError
+
+    @abstractmethod
+    def model_to_device(self) -> None:
+        """Moves the model to the correct device"""
+        raise NotImplementedError
+
+    @property
+    @abstractmethod
+    def is_global_zero(self) -> bool:
+        """Whether the current process is the rank zero process not only on the local node, but for all nodes."""
+        raise NotImplementedError
+
+    @abstractmethod
+    def reduce(self, output: Union[torch.Tensor, Any], *args: Any, **kwargs: Any) -> Union[torch.Tensor, Any]:
+        """Reduces the given output (e.g. across GPUs/Processes)"""
+        raise NotImplementedError
+
+    @abstractmethod
+    def barrier(self, name: Optional[str] = None) -> None:
+        """Forces all possibly joined processes to wait for each other"""
+        raise NotImplementedError
+
+    @abstractmethod
+    def broadcast(self, obj: object, src: int = 0) -> object:
+        """Broadcasts an object to all processes"""
+        raise NotImplementedError
+
+    # TODO method this is currently unused. Check after complete refactors are pushed
+    def set_nvidia_flags(self, is_slurm_managing_tasks: bool, device_ids: Optional[Sequence]) -> None:
+        if device_ids is None:
+            return
+
+        # set the correct cuda visible devices (using pci order)
+        os.environ["CUDA_DEVICE_ORDER"] = "PCI_BUS_ID"
+        all_gpu_ids = ",".join([str(x) for x in range(torch.cuda.device_count())])
+        devices = os.environ.get("CUDA_VISIBLE_DEVICES", all_gpu_ids)
+        if self.lightning_module is not None:
+            log.info(f"LOCAL_RANK: {self.lightning_module.trainer.local_rank} - CUDA_VISIBLE_DEVICES: [{devices}]")
```

Comment:
why's this a part of the base training type plugin? should this go in some utils instead?

Reply:
It is there because it was before. I think this is some legacy code and will be removed completely (as indicated by the todo). Just need to double check when everything is merged.